sockeye/inference.py

```diff
@@ -1690,37 +1690,51 @@ def _beam_search(self,
-                                                                       models_output_layer_b=models_output_layer_b)
-
-            # (2) Update scores. Special treatment for finished and inactive rows. Inactive rows are inf everywhere;
+            target_dists, attention_scores, model_states = self._decode_step(prev_word=best_word_indices,
+                                                                             step=t,
+                                                                             source_length=source_length,
+                                                                             states=model_states,
+                                                                             models_output_layer_w=models_output_layer_w,
+                                                                             models_output_layer_b=models_output_layer_b)
+
+            # (2) Produces the accumulated cost of target words in each row.
+            # There is special treatment for finished and inactive rows: inactive rows are inf everywhere;
             # finished rows are inf everywhere except column zero, which holds the accumulated model score
-            scores = self._update_scores.forward(scores, finished, inactive, scores_accumulated, self.inf_array, pad_dist)
+            scores = self._update_scores.forward(target_dists, finished, inactive, scores_accumulated, pad_dist)
 
             # Mark entries that should be blocked as having a score of np.inf
             if self.global_avoid_trie or any(raw_avoid_list):
                 block_indices = avoid_states.avoid()
                 if len(block_indices) > 0:
                     scores[block_indices] = np.inf
 
+            # finished_on_prev_step = (best_word_indices == self.vocab_target[C.EOS_SYMBOL]) + (best_word_indices == C.PAD_ID)
+
             # (3) Get beam_size winning hypotheses for each sentence block separately. Only look as
             # far as the active beam size for each sentence.
-            best_hyp_indices, best_word_indices, scores_accumulated = self._top(scores)
+
+            # When sampling, manually set the chosen word to C.PAD_ID for finished hypotheses
+            if self.sample:
+                best_hyp_indices, best_word_indices, scores_accumulated = self._top(scores, target_dists, finished)
+            else:
+                # On the first timestep, all hypotheses have identical histories, so force topk() to choose extensions
+                # of the first row only
+                if t == 1 and not self.skip_topk:
+                    batch_indices = mx.nd.array(np.arange(0, self.batch_size * self.beam_size, self.beam_size), dtype='int32', ctx=self.context)
+                    best_hyp_indices, best_word_indices, scores_accumulated = self._top(scores[batch_indices,:])
+                else:
+                    best_hyp_indices, best_word_indices, scores_accumulated = self._top(scores)
 
```

Comment:
This restores the t==1 special case. We pass only the first row of `scores` (for all items in the batch).

Reply:
For brevity, could we do 
```python
if t == 1 and not self.skip_topk:
  scores = scores[self.batch_indices, :]
best_hyp_indices, best_word_indices, scores_accumulated = self._top(scores)
```
, or does this interfere with line 1746 (passing unmodified `scores` into `constrained.topk`? If yes, can we change the constraint code to work with this simplification?